infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -129,10 +131,37 @@ public LogUnitServer(ServerContext serverContext) {
     /**
      * Service an incoming request for maximum global address the log unit server has written.
      */
-    @ServerHandler(type = CorfuMsgType.TAIL_REQUEST)
-    public void handleTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+    @ServerHandler(type = CorfuMsgType.LOG_TAIL_REQUEST)
+    public void handleLogTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        Long logTail = batchWriter.queryLogTail(msg.getEpoch());
+        r.sendResponse(ctx, msg, CorfuMsgType.LOG_TAIL_RESPONSE.payloadMsg(logTail));
+    }
+
+    /**
+     * Service an incoming request for all stream tails (also returns log global tail).
+     */
+    @ServerHandler(type = CorfuMsgType.TAILS_REQUEST)
+    public void handleTailsRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
         TailsResponse tails = batchWriter.queryTails(msg.getEpoch());
-        r.sendResponse(ctx, msg, CorfuMsgType.TAIL_RESPONSE.payloadMsg(tails));
+        r.sendResponse(ctx, msg, CorfuMsgType.TAILS_RESPONSE.payloadMsg(tails));
+    }
+
+    /**
+     * Service an incoming request for streams addresses, i.e., the map of addresses belonging to each stream.
+     */
+    @ServerHandler(type = CorfuMsgType.ALL_STREAMS_ADDRESS_REQUEST)
```

Comment:
Maybe this is fine now, but stream address space management should be thread safe so that we dont need to serialize all the readers.

Reply:
As per our conversation the other day, I'll leave it for now serialized through the batchProcessor as It is not thread safe. Also this is also only called on sequencer recovery, if we ever plan to balance address requests between sequencer and log units, it is important that we do not serialize readers.